cpp/src/arrow/io/memory.h

```diff
@@ -146,12 +146,9 @@ class ARROW_EXPORT BufferReader
     : public internal::RandomAccessFileConcurrencyWrapper<BufferReader> {
  public:
   explicit BufferReader(std::shared_ptr<Buffer> buffer);
-  explicit BufferReader(const Buffer& buffer);
-  BufferReader(const uint8_t* data, int64_t size);
 
-  /// \brief Instantiate from std::string or std::string_view. Does not
-  /// own data
-  explicit BufferReader(std::string_view data);
```

Comment:
Maybe we should deprecate instead of remove?

Reply:
We could even introduce static factories that have the same behavior as these constructors, but whose names explicitly indicate that they may be unsafe

Reply:
OK, seems that I should mark them deprecated, but keep them as before (including zero-copy)

Then how do I fix https://github.com/apache/arrow/pull/37271#issuecomment-1686585639 ? Using a `BufferReader::FromString`? Or other way?

Reply:
Ah, that's annoying.

`FromString` is probably best.